test/core/security/credentials_test.cc

```diff
@@ -1512,16 +1512,105 @@ static void test_no_google_default_creds(void) {
+  GPR_ASSERT(grpc_google_default_credentials_create(nullptr) == nullptr);
   GPR_ASSERT(g_test_gce_tenancy_checker_called == true);
   /* Cleanup. */
   grpc_override_well_known_credentials_path_getter(nullptr);
   grpc_httpcli_set_override(nullptr, nullptr);
 }
 
+static void test_google_default_creds_call_creds_specified(void) {
+  expected_md emd[] = {
+      {"authorization", "Bearer ya29.AHES6ZRN3-HlhAPya30GnW_bHSb_"}};
+  request_metadata_state* state =
+      make_request_metadata_state(GRPC_ERROR_NONE, emd, GPR_ARRAY_SIZE(emd));
+  grpc_auth_metadata_context auth_md_ctx = {test_service_url, test_method,
+                                            nullptr, nullptr};
+  grpc_core::ExecCtx exec_ctx;
+  grpc_flush_cached_google_default_credentials();
+  grpc_call_credentials* call_creds =
+      grpc_google_compute_engine_credentials_create(nullptr);
+  set_gce_tenancy_checker_for_testing(test_gce_tenancy_checker);
+  g_test_gce_tenancy_checker_called = false;
+  g_test_is_on_gce = true;
+  grpc_httpcli_set_override(
+      default_creds_metadata_server_detection_httpcli_get_success_override,
+      httpcli_post_should_not_be_called);
+  grpc_composite_channel_credentials* channel_creds =
+      reinterpret_cast<grpc_composite_channel_credentials*>(
+          grpc_google_default_credentials_create(call_creds));
+  GPR_ASSERT(g_test_gce_tenancy_checker_called == true);
+  GPR_ASSERT(channel_creds != nullptr);
+  GPR_ASSERT(channel_creds->call_creds() != nullptr);
+  grpc_httpcli_set_override(compute_engine_httpcli_get_success_override,
+                            httpcli_post_should_not_be_called);
+  run_request_metadata_test(channel_creds->mutable_call_creds(), auth_md_ctx,
+                            state);
+  grpc_core::ExecCtx::Get()->Flush();
+  channel_creds->Unref();
+  grpc_httpcli_set_override(nullptr, nullptr);
+}
+
+struct fake_call_creds : public grpc_call_credentials {
+ public:
+  explicit fake_call_creds() : grpc_call_credentials("fake") {
+    grpc_slice key = grpc_slice_from_static_string("foo");
+    grpc_slice value = grpc_slice_from_static_string("oof");
+    dummy_md_ = grpc_mdelem_from_slices(key, value);
+    grpc_slice_unref(key);
+    grpc_slice_unref(value);
+  }
+
+  ~fake_call_creds() { GRPC_MDELEM_UNREF(dummy_md_); }
+
+  bool get_request_metadata(grpc_polling_entity* pollent,
+                            grpc_auth_metadata_context context,
+                            grpc_credentials_mdelem_array* md_array,
+                            grpc_closure* on_request_metadata,
+                            grpc_error** error) {
+    grpc_credentials_mdelem_array_add(md_array, dummy_md_);
+    return true;
+  }
+
+  void cancel_get_request_metadata(grpc_credentials_mdelem_array* md_array,
+                                   grpc_error* error) {}
+
+ private:
+  grpc_mdelem dummy_md_;
+};
+
+static void test_google_default_creds_not_default(void) {
+  expected_md emd[] = {{"foo", "oof"}};
```

Comment:
Oof!  I feel like I've just been punched in the gut! :)

Reply:
🙂 https://www.youtube.com/watch?v=Ms5yoA1DvH0